src/geb/geb.lisp

```diff
@@ -157,3 +157,49 @@ Example:
         ((distribute x y z)  (coprod (prod x y) (prod x z)))
         ((inject-left x y)   (coprod x y))
         ((inject-right x y)  (coprod x y)))))
+
+(defun gather (x y z)
+  (pair (mcase (<-left x y) (<-left x z))
+        (mcase (comp (->left y z) (<-right x y))
+               (comp (->right y z) (<-right x z)))))
+
+(defun prod-left-assoc (f)
+  (trivia:match (dom f)
+    ((prod w (prod x y))
+     (comp f (pair (comp (<-left w x) (<-left (prod w x) y))
+                   (pair (comp (<-right w x) (<-left (prod w x) y))
+                         (<-right (prod w x) y)))))
+    (_ (error "object ~A need to be of a double product type, however it is of ~A" f (dom f)))))
+
+(defun right (f)
+  (trivia:match (dom f)
+    ((coprod x y)
+     (comp f (->right x y)))
+    (_ (error "object ~A need to be of a coproduct type, however it is of ~A"
+              f (dom f)))))
+
+(defun left (f)
+  (trivia:match (dom f)
+    ((coprod x y)
+     (comp f (->left x y)))
+    (_ (error "object ~A need to be of a coproduct type, however it is of ~A"
+              f (dom f)))))
+
+(defgeneric curry (f)
+  (:documentation "Curries the given object"))
+
+(defmethod curry ((f <substmorph>))
+  (cond ((not (typep f 'substmorph))
+         (subclass-responsibility f))
+        ((not (typep (dom f) 'prod))
+         (error "object ~A need to be of a product type, however it is of ~A" f (dom f)))
+        (t
+         (let ((dom (dom f)))
+           (match-of substobj (mcadr dom)
+             (alias        (curry f))
```

Comment:
Does this not make an infinite loop if `(mcadr (dom f))` is an alias like `bool`?

I suspect you'll want to [`labels` ](http://www.lispworks.com/documentation/lw70/CLHS/Body/s_flet_.htm) the `match-of` and call it recursively in the alias case